test/e2e/service_test.go

```diff
@@ -81,3 +82,60 @@ func (data *TestData) createClusterIPServiceAndBackendPods(t *testing.T, name st
 
 	return svc, cleanup
 }
+
+// TestNodePortWindows tests NodePort Service on Windows Node. It is a temporary test to replace upstream Kubernetes one:
+// https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/test/e2e/windows/service.go#L42
+func TestNodePortWindows(t *testing.T) {
+	skipIfNoWindowsNodes(t)
+
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	svcName := "agnhost"
+	if len(clusterInfo.windowsNodes) == 0 {
+		t.Fatalf("The cluster should have at least one Windows Node")
+	}
+	svcNode := nodeName(clusterInfo.windowsNodes[0])
+	svc, cleanup := data.createAgnhostServiceAndBackendPods(t, svcName, svcNode, corev1.ServiceTypeNodePort)
+	defer cleanup()
+	t.Logf("%s Service is ready", svcName)
+
+	clientName := "agnhost-client"
+	require.NoError(t, data.createAgnhostPodOnNode(clientName, nodeName(0)))
+	defer data.deletePodAndWait(defaultTimeout, clientName)
+	_, err = data.podWaitForIPs(defaultTimeout, clientName, testNamespace)
+	require.NoError(t, err)
+
+	nodeIP := clusterInfo.nodes[0].ip
+	nodePort := int(svc.Spec.Ports[0].NodePort)
+	addr := fmt.Sprintf("http://%s:%d", nodeIP, nodePort)
+
+	cmd := append([]string{"curl", "--connect-timeout", "1", "--retry", "5", "--retry-connrefused"}, addr)
+	stdout, stderr, err := data.runCommandFromPod(testNamespace, clientName, agnhostContainerName, cmd)
+	if err != nil {
+		t.Errorf("Error when running command '%s' from Pod '%s', stdout: %s, stderr: %s, error: %v",
+			strings.Join(cmd, " "), clientName, stdout, stderr, err)
+	} else {
+		t.Logf("curl from Pod '%s' to `%s` succeeded", clientName, addr)
+	}
+}
+
+func (data *TestData) createAgnhostServiceAndBackendPods(t *testing.T, name string, node string, svcType corev1.ServiceType) (*corev1.Service, func()) {
+	ipv4Protocol := corev1.IPv4Protocol
+	require.NoError(t, data.createAgnhostServicePodOnNode(name, node))
+	_, err := data.podWaitForIPs(defaultTimeout, name, testNamespace)
+	require.NoError(t, err)
+	require.NoError(t, data.podWaitForRunning(defaultTimeout, name, testNamespace))
+	svc, err := data.createAgnhostService(name, false, &ipv4Protocol, svcType)
```

Comment:
it's not clear to me why we need helper functions `createAgnhostServicePodOnNode` and `createAgnhostService`, as they are themselves wrappers around other methods. It feels like we could take care of everything in this function directly, unless we need these helper functions for something else.

Reply:
The reason is that I saw same usage for nginx service/pod.
https://github.com/antrea-io/antrea/blob/main/test/e2e/framework.go#L961
https://github.com/antrea-io/antrea/blob/main/test/e2e/framework.go#L1369

However, I can directly use those wrapped methods this time since it is the only place they are used. In the future, if such methods are used more, I think we can refactor even merge the nginx and agnhost methods into one?